docs/api/synchronization.rst

```diff
@@ -0,0 +1,160 @@
+#. If response is ``200 OK`` → repeat previous step.
+#. If response is ``412 Precondition Failed``, some changes occured since the last
+   page → Store the ``ETag``response header into ``before``
+#. Fetch and handle the changes using ``GET collections?_since=<timestamp>&_before=<before>``,
+   and store the ``ETag``response header in to ``timestamp``
+#. Go back to step 5 (follow the ``Next-Page``)
+
+
+Apply changes
+=============
+
+Applying remote changes to the local database consists in adding new records,
+updating changed records and remove deleted records.
+
+From the client perspective, *Kinto* does not distinguish creations from updates.
+In the *polling for changes* response, created records are simply the records
+unknown by the client (using ``id`` field).
+
+If the records to be updated or deleted had also been modified locally then
+the developper must choose a relevant strategy. For example, merge fields or
+ignore deletion.
+
+
+Concurrency control
+===================
+
+As described in :ref:`server-timestamps`, *Kinto* uses *ETag* for concurrency
+control.
+
+ETags are provided in response headers, for the collection as well as individual
+records.
+
+Even though it is recommended to consider them as opaque and abstract, it can still
+be useful to notice that ETags are a string with the quoted record last modified value
+(``"<record.last_modified>"``)
+
+
+Protected creation with PUT
+---------------------------
+
+Add a ``If-None-Match: *`` request header to the ``PUT`` to make sure no
+record exists on the server with this ID.
+
+This can be useful to avoid overwrites when creating records with ``PUT``
+instead of ``POST``.
+
+
+Protected update and delete
+---------------------------
+
+Add a ``If-Match: "<record.last_modified>"`` request header to the ``PUT``, ``PATCH``
+or ``DELETE`` request.
+
+*Kinto* will reject the request with a ``412 Precondition Failed`` response if
+the record was modified in the interim.
+
+If the remote record was already deleted, a ``404 Not found`` response will be
+returned. The client can choose to ignore it.
+
+
+Offline-first
+=============
+
+Since the server won't be available to assign record identifiers while offline,
+it is recommended to generate them on the client.
+
+Record identifiers are `UUID <https://en.wikipedia.org/wiki/Universally_unique_identifier>`_,
+a very common format for unique strings with almost zero [#]_ collision probability.
+
+When going back online, the set of changes can be sent to the server using a
+:ref:`batch` request.
+
+
+.. _sync-implementations:
+
+Implementations
+===============
+
+The **current implementation of reference** for offline-first records synchronization is
+:rtd:`Kinto.js <kintojs>`_.
+
+
+Before that, some other clients were implemented in the context of the
+*ReadingList* project, such as `RL Web client`_ (React.js), `Android RL sync`_ (Java) or `Firefox RL client`_ (asm.js).
```

Comment:
[Oxford Comma](https://richardgilbert.wordpress.com/2012/06/22/we-like-the-oxford-comma-ii/)